### DISCUSSION TITLE
Directional Gravity

### DISCUSSION THREAD

**Discussion by Stonerhino3D** [Sixze/ALS-Refactored#422] [Ideas]
I noticed that in 5.3 they added all the code for directional gravity. They then greyed it out in the editor. It can be turned on for custom movement classes like ALS as I understand it. Does anyone here with c++ talent know how to add this. With all the walkable planets coming in unreal I think it would be a huge popular feature. Would also jump start people using World Scape. Or Voxel planets. ALS Refactored could be drop in for planets. https://forums.unrealengine.com/t/5-3-default-character-movement-controller-gravity-direction-greyed-out/1356641/2  The adjustments needed are in this post. I just don't know anything about c++ but it looks like a small addition.

**Comment by ameaninglessname** [comment]
These changes are only need for blueprint project (it just exposed the `SetGravityDirection` interface to BP), for cpp project, it's already there.
(And the `UFUNCTION` macro should be better added to engine code IMO)

**Comment by Stonerhino3D** [comment]
I was referring to those of us that are using blueprint only.  Like myself. I would love to have SetGravityDirection exposed in blueprints. And since the changes can as I understand it be implemented on custom character movement classes like ALS. Those of us that don't know how to make a new custom character class and recompile the engine code can just use the custom gravity for their character. Most of the plugins and prior fixes like Ninja Character have shut down because the gravity direction was added to the code just not exposed in blueprint or editor. If someone could help with instructions for editing the ALS Refactored code to expose SetGravityDirection to blueprint I would be ever so grateful.

**Reply by ameaninglessname** [reply]
You could add these code to `UAlsCharacterMovementComponent` in `AlsCharacterMovementComponent.h`,

```cpp
	UFUNCTION(BlueprintCallable)
	virtual void SetGravityDirection(const FVector& GravityDir) override
	{
		Super::SetGravityDirection(GravityDir);
	}
```

then you could use it in blueprint.

**Reply by Stonerhino3D** [reply]
What line would you add that to? I added it a couple of places but it gave errors.

**Reply by ameaninglessname** [reply]
![image](https://github.com/Sixze/ALS-Refactored/assets/20263103/523de2db-c92f-44e1-81bb-6b5ecc167d44)

**Reply by Stonerhino3D** [reply]
It wanted a category to compile. I tried adding a dummy category. 
	UFUNCTION(BlueprintCallable, Category = "")
	virtual void SetGravityDirection(const FVector& GravityDir) override
	{
		Super::SetGravityDirection(GravityDir);
	}

also tried

	UFUNCTION(BlueprintCallable, Category = "ALS|Character Movement")
	virtual void SetGravityDirection(const FVector& GravityDir) override
	{
		Super::SetGravityDirection(GravityDir);
	}

Compiled. But I cant seem to call the Blueprint. I tried casting to different ALS cast's but I cant find Set GravityDirection in the editor. Maybe I have the wrong category.

**Reply by ameaninglessname** [reply]
Me, from `B_Als_Character`
![image](https://github.com/Sixze/ALS-Refactored/assets/20263103/8d02dcdf-e154-4ea4-bd38-79a256caca55)

**Reply by Stonerhino3D** [reply]
TY so much for your help. I got it working following all your steps getting a new copy of 4.13 and adding it to a new project. The engine asked to rebuild the ALS components. I said yes and booted right in. I setup the blueprint code to change the gravity towards planet centre. Its working perfectly running around a small sphere. I absolutely love it.

I need to learn how to force the engine to rebuild my current ALS in my main project like this one did from scratch. I thought running the BuildPlugin.bat was rebuilding/compiling.

**Reply by Stonerhino3D** [reply]
I managed to get the top. Half on the sphere working amazing no issues. Now comes the bottom half. I suspect the maximum floor angle at 90 turns the ALS character into fall mode for the bottom half of the globe.  Any quick ideas on where I would stop the transition.
